enclave/sgx/qeidinfo.c

```diff
@@ -8,96 +8,103 @@
 {
     oe_result_t result = OE_FAILURE;
-    size_t host_args_buffer_size = sizeof(*args);
-    uint8_t* host_args_buffer = NULL;
-    oe_get_qe_identity_info_args_t* host_args = NULL;
-    oe_get_qe_identity_info_args_t tmp_args = {0};
+    const size_t QE_ID_INFO_SIZE = OE_PAGE_SIZE;
+    const size_t ISSUER_CHAIN_SIZE = OE_PAGE_SIZE;
+    uint32_t retval;
+    oe_get_qe_identity_info_args_t args;
 
-    if (args == NULL)
+    if (args_out == NULL)
         OE_RAISE(OE_FAILURE);
 
-    // allocate host memory to hold enclave arguments before calling into host
-    // for getting qe identity info
-    host_args_buffer = oe_host_malloc(host_args_buffer_size);
-    if (host_args_buffer == NULL)
+    memset(args_out, 0, sizeof(oe_get_qe_identity_info_args_t));
+    memset(&args, 0, sizeof(oe_get_qe_identity_info_args_t));
+
+    if (!(args.qe_id_info = malloc(QE_ID_INFO_SIZE)))
         OE_RAISE(OE_OUT_OF_MEMORY);
 
-    // Copy args struct.
-    host_args = (oe_get_qe_identity_info_args_t*)host_args_buffer;
-    *host_args = *args;
-
-    result = oe_ocall(OE_OCALL_GET_QE_ID_INFO, (uint64_t)host_args, NULL);
-    OE_CHECK(result);
-
-    // Copy args to prevent TOCTOU issues.
-    tmp_args = *host_args;
-    if (tmp_args.result == OE_QUOTE_PROVIDER_CALL_ERROR)
-        OE_TRACE_WARNING(
-            "Warning: QE Identity was not supported by quote provider\n");
-
-    OE_CHECK_NO_TRACE(tmp_args.result);
-
-    if (tmp_args.host_out_buffer == NULL ||
-        !oe_is_outside_enclave(tmp_args.host_out_buffer, sizeof(uint8_t)))
-        OE_RAISE(OE_UNEXPECTED);
-
-    // Copy the return data back into enclave address space
-    // Ensure that all required outputs exist.
-    OE_CHECK(_copy_buffer_to_enclave(
-        &args->qe_id_info,
-        &args->qe_id_info_size,
-        tmp_args.qe_id_info,
-        tmp_args.qe_id_info_size));
-    OE_CHECK(_copy_buffer_to_enclave(
-        &args->issuer_chain,
-        &args->issuer_chain_size,
-        tmp_args.issuer_chain,
-        tmp_args.issuer_chain_size));
+    if (!(args.issuer_chain = malloc(ISSUER_CHAIN_SIZE)))
+        OE_RAISE(OE_OUT_OF_MEMORY);
+
+    args.qe_id_info_size = QE_ID_INFO_SIZE;
+    args.issuer_chain_size = ISSUER_CHAIN_SIZE;
+
+    /* First call (one or more buffers might be too small). */
+    if (oe_get_qe_identify_info_ocall(
+            &retval,
+            args.qe_id_info,
+            args.qe_id_info_size,
+            &args.qe_id_info_size,
+            args.issuer_chain,
+            args.issuer_chain_size,
+            &args.issuer_chain_size) != OE_OK)
+    {
+        OE_RAISE(OE_FAILURE);
+    }
+
+    /* Subsequent calls to expand the buffers as needed. Note that the output
+     * buffer sizes returned from the first call might still be too small on
+     * the next call since the qe_id_info and issuer_chain could change
+     * between calls. To compensate, loop until success.
+     */
+    while ((oe_result_t)retval == OE_BUFFER_TOO_SMALL)
```

Comment:
Should we give up after n iterations?
What are the security implications of looping?

Reply:
Yeah, I chose 8 for now. This is a very unlikely scenario but the certificate source could change at the exact time of call but it is a virtual impossibility that it will change twice in under a few milliseconds.

I don't think the security implementations are different than looping just once, since in any iteration the host has complete control over the returned data.